gen/version.go

```diff
@@ -0,0 +1,185 @@
+	}
+
+	fmt.Println("Found middleware version", middlewareVersion)
+
+	tmpl := template.Must(template.New("version.go").Parse(versionFileTemplate))
+
+	versionFile, err := os.Create(rosettaVersionFilePath)
+	if err != nil {
+		log.Fatalf("could not open version file: %v", err)
+	}
+
+	args := struct {
+		RosettaVersion    string
+		NodeVersion       string
+		MiddlewareVersion string
+	}{
+		RosettaVersion:    rosettaVersion,
+		NodeVersion:       nodeVersion,
+		MiddlewareVersion: middlewareVersion,
+	}
+
+	err = tmpl.Execute(versionFile, args)
+	if err != nil {
+		log.Fatalf("could not execute template: %v", err)
+	}
+}
+
+func NodeVersion() (string, error) {
+	// Fetch Node version from the go.mod file.
+	gomod, err := os.ReadFile("../go.mod")
+	if err != nil {
+		return "", fmt.Errorf("could not read go mod file: %v", err)
+	}
+
+	modfile, err := modfile.Parse("go.mod", gomod, nil)
+	if err != nil {
+		return "", fmt.Errorf("could not parse go mod file: %v", err)
+	}
+
+	var nodeVersion string
+	for _, module := range modfile.Require {
+		if module.Mod.Path == "github.com/onflow/flow-go" {
+			nodeVersion = module.Mod.Version
+			break
+		}
+	}
+
+	if nodeVersion == "" {
+		return "", fmt.Errorf("could not find github.com/onflow/flow-go dependency in go mod file")
+	}
+
+	return nodeVersion, nil
+}
+
+func MiddlewareVersion() (string, error) {
+	// Fetch middleware version by looking at the latest tag on the repository.
+	repo, err := git.PlainOpen("../")
+	if err != nil {
+		return "", fmt.Errorf("unable to open local git repository: %w", err)
+	}
+
+	tags, err := repo.Tags()
+	if err != nil {
+		return "", fmt.Errorf("unable to find local git tags: %w", err)
+	}
+
+	// Fetch all tags and which commit they reference.
+	tagsMap := make(map[plumbing.Hash]*plumbing.Reference)
+	err = tags.ForEach(func(t *plumbing.Reference) error {
+		tagsMap[t.Hash()] = t
+		return nil
+	})
+
+	head, err := repo.Head()
+	if err != nil {
+		return "", fmt.Errorf("unable to find local git HEAD: %w", err)
+	}
+
+	// Fetch the reference log.
+	cIter, err := repo.Log(&git.LogOptions{
+		From:  head.Hash(),
+		Order: git.LogOrderCommitterTime,
+	})
+	if err != nil {
+		return "", fmt.Errorf("unable to read local git log: %w", err)
+	}
+
+	// Search for the latest tag on the current branch.
+	var tag *plumbing.Reference
+	var count int
+	err = cIter.ForEach(func(c *object.Commit) error {
+		if t, ok := tagsMap[c.Hash]; ok {
+			tag = t
+		}
+		if tag != nil {
+			return storer.ErrStop
+		}
+		count++
+		return nil
+	})
+
+	// Repository does not have any tags, return placeholder.
+	if tag == nil {
+		return "v0.0.0", nil
+	}
+
+	// If the current branch was just tagged, the version is precisely this tag.
+	if count == 0 {
+		return fmt.Sprint(tag.Name().Short()), nil
+	}
+
+	// Otherwise, generate a version name from the latest tag name with HEAD's commit hash.
+	tagName := fmt.Sprintf("%v-%v", tag.Name().Short(), head.Hash())
```

Comment:
I don't think we should use the tag with the commit hash behind it. In Go modules, this is used as a convention to reference a specific tag, as far as I'm aware, and not a commit after that tag, so this could be confusing. Is this taken from some kind of convention?

Maybe we could use `branch-commit` instead?

Reply:
I am fine with `branch-commit`, but the reason why I usually like the `latesttag-commit` is that it usually gives you an idea of the actual features/breaking changes to expect. For example, when we work with Flow-Go dependencies, assuming we would use their private builds, imagine that instead of versions being named `0.15.1-patchX` they were named `add-X-feature-545d489sac4` :p 

Having the latest tag makes all the difference IMO, as you can know if it's compatible with what you're doing without having to check for yourself at which tag that branch is/was at the given commit.

If you both strongly think having branch-commit it better though, I'm ok with changing it.